fix: incorporate block assembler to event queue

dwebchapey opened the pull request:
This PR fixes #173 

1.  Enqueue `conditionalMakeBlock()` to `fastQueue` via `queueBlockAssembler()` . This prevents `makeBlock()` from executing while on chain events are getting processed.
2.  `queueBlockAssembler()` is triggered on two scenarios 

-  from `newCurrentProposerEventHandler` when proposer is me
- after completion of `conditionalMakeBlock()` given proposer is still me , priority zero queue is zero and enough unprocessed transactions are available in local db

Before calling `makeBlocks()` , we do 2 checks

1. proposer.IsMe  ?  move to stage 2  : exit without re-queuing. This stops putting assemblers into the queue when we lose the proposer role
2. fastQueue === 0 && unprocessedTx >= txPerBlock ? proceed to makeBlocks , re-enqueue event  and exit: re-queue event and exit.

Before  re-enqueue , wait for `fastQueue` to be empty and also db to procure enough unprocessed transactions . This will prevent unnecessary en-queue and de-queue operations. 
The idea is that we fail fast if any of these checks fail, and we optimistically re-enqueue the `conditionalMakeBlock` also reducing queue operations.

IlyasRidhuan commented on nightfall-optimist/src/services/event-queue.mjs:
```diff
@@ -17,21 +17,23 @@ Web3js handles event removal (due to a chain reorg) by re-emitting the (now remo
 event with its `removed` property set to true. In the code below, we look out for
 and catch these removals, processing them appropriately.
 */
+/* eslint-disable import/no-cycle */
```
Cyclic imports are an anti-pattern and so we should not just disable the check if one exists.

The specific block-assemble logic should be kept out of this file. There is a strong chance this event-queue code will need to re-used in `client` and thus elevated to a `common-files` level.

Instead a generic enqueue function would be preferred - this should also prevent the cyclic dependency

IlyasRidhuan commented on nightfall-optimist/src/services/event-queue.mjs:
```diff
@@ -43,6 +45,19 @@ export function nextHigherPriorityQueueHasEmptied(priority) {
+*/
+export async function queueBlockAssembler(proposer) {
+  queues[0].push(async () => {
+    await nextHigherPriorityQueueHasEmptied(0); // prevent conditionalmakeblock from running until fastQueue is emptied
```
`await` called on queue 0 has no impact , so i'm unclear why it is being called here. We should not need to do a wait on empty check here

dwebchapey replied:
`priority` is passed as an argument to make function generic to all queues

IlyasRidhuan commented on nightfall-optimist/src/services/block-assembler.mjs:
```diff
@@ -41,18 +42,23 @@ async function makeBlock(proposer, number = TRANSACTIONS_PER_BLOCK) {
+    // it is important for queues length to be empty since a rollback or any onchain event following
+    // in queue after blockassemble should not be done So if other events are present in queue
+    // we wait for queue to empty and re enqueue blockassembler again
+    makeBlocks = false;
```
I don't see the effect that `makeBlocks` has on controlling the number of `conditionalMakeBlock` instances are in the queue.

dwebchapey commented:
- Removed cyclic dependency by creating a generic enqueue function `eventQueueManager`
- Parametrised `eventQueueManager` with callback , priority and args 
- Removed dependency on `makeBlocks` for single instance check

IlyasRidhuan commented on nightfall-optimist/src/services/event-queue.mjs:
```diff
@@ -43,6 +43,13 @@ export function nextHigherPriorityQueueHasEmptied(priority) {
 
+export async function eventQueueManager(callback, priority, args) {
+  queues[priority].push(async () => {
+    await nextHigherPriorityQueueHasEmptied(priority); // prevent conditionalmakeblock from running until fastQueue is emptied
```
I'm unclear with what this line does

IlyasRidhuan commented on nightfall-optimist/src/services/block-assembler.mjs:
```diff
@@ -41,18 +41,18 @@ async function makeBlock(proposer, number = TRANSACTIONS_PER_BLOCK) {
+  // to be processed, we can assemble a block and create a proposal
+  // transaction. If not, we must wait until either we have enough (hooray)
+  // or we're no-longer the proposer (boo).
+  if (proposer.isMe) {
```
Is there a reason this is split across two `if` statements?

dwebchapey replied:
The outer `If` does not re-enqueue `conditonalMakeBlock` since proposer is no longer the host node.

IlyasRidhuan commented on nightfall-optimist/src/services/block-assembler.mjs:
```diff
@@ -76,10 +76,11 @@ export async function conditionalMakeBlock(proposer) {
+      await eventQueueManager(conditionalMakeBlock, 0, proposer);
     }
-    // slow the loop a bit so we don't hammer the database
-    await new Promise(resolve => setTimeout(resolve, 1000));
```
Do we still need some slowing down of this function call?

IlyasRidhuan replied:
@dwebchapey
